### PR TITLE
[Paddle-TRT] YoloV3 OPs on TensorRT

### PR DESCRIPTION
**Pull request by zlsh80826**
### PR types

New features

### PR changes

OPs

### Describe

Add ``multiclass_nms`` and ``nearest_interpret`` tensorrt converter

### CODE REVIEW

File: paddle/fluid/inference/tensorrt/convert/CMakeLists.txt [hunk 1]
```diff
@@ -5,6 +5,7 @@ nv_library(tensorrt_converter
                 pad_op.cc split_op.cc prelu_op.cc leaky_relu_op.cc gelu_op.cc layer_norm_op.cc multihead_matmul_op.cc
                 shuffle_channel_op.cc swish_op.cc instance_norm_op.cc stack_op.cc
                 emb_eltwise_layernorm.cc skip_layernorm.cc scale_op.cc slice_op.cc hard_sigmoid_op.cc hard_swish_op.cc clip_op.cc
+                nearest_interp.cc multiclass_nms.cc
```
Comment (cryoco) [inline paddle/fluid/inference/tensorrt/convert/CMakeLists.txt]: 和其他op converter文件统一下文件名为nearest_interp_op.cc multiclass_nms_op.cc吧

File: paddle/fluid/inference/tensorrt/convert/multiclass_nms.cc [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+}  // namespace proto
+}  // namespace framework
+}  // namespace paddle
+
+namespace paddle {
+namespace inference {
+namespace tensorrt {
+
+/*
+ * ConcatOp
```
Comment (cryoco) [inline paddle/fluid/inference/tensorrt/convert/multiclass_nms.cc]: 改为对应op名

File: paddle/fluid/inference/tensorrt/convert/nearest_interp.cc [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+}  // namespace proto
+}  // namespace framework
+}  // namespace paddle
+
+namespace paddle {
+namespace inference {
+namespace tensorrt {
+
+/*
+ * ConcatOp
```
Comment (cryoco) [inline paddle/fluid/inference/tensorrt/convert/nearest_interp.cc]: 改为对应op名

File: paddle/fluid/inference/tensorrt/convert/multiclass_nms.cc [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+ */
+
+void show_dims(nvinfer1::ITensor* tensor, const std::string& name) {
+  auto dims = tensor->getDimensions();
+  std::cerr << name << ": ";
+  std::cerr << dims.nbDims << ", ";
+  for (int i = 0; i < dims.nbDims; ++i) {
+    std::cerr << dims.d[i] << " ";
+  }
+  std::cerr << std::endl;
```
Comment (cryoco) [inline paddle/fluid/inference/tensorrt/convert/multiclass_nms.cc]: 用GLOG的方式来打log吧，可以用VLOG(1)，因为这个dim信息并非是必须输出的

File: paddle/fluid/inference/tensorrt/convert/multiclass_nms.cc [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+    int background_label =
+        BOOST_GET_CONST(int, op_desc.GetAttr("background_label"));
+    float score_threshold =
+        BOOST_GET_CONST(float, op_desc.GetAttr("score_threshold"));
+    int nms_top_k = BOOST_GET_CONST(int, op_desc.GetAttr("nms_top_k"));
+    float nms_threshold =
+        BOOST_GET_CONST(float, op_desc.GetAttr("nms_threshold"));
+    int keep_top_k = BOOST_GET_CONST(int, op_desc.GetAttr("keep_top_k"));
+    bool normalized = BOOST_GET_CONST(bool, op_desc.GetAttr("normalized"));
```
Comment (cryoco) [inline paddle/fluid/inference/tensorrt/convert/multiclass_nms.cc]: 这些Attr在multiclass_nms op中不一定都有，可以先用op_desc.HasAttr(attr)的方式检查下，没有的话设为默认值

**Review by cryoco — COMMENTED** [review]
建议在python/paddle/fluid/test/unittests/ir/inference/下添加对应trt op converter的python单测，参考已有的其他单测如[test_trt_subgraph_pass.py](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/tests/unittests/ir/inference/test_trt_subgraph_pass.py)即可

File: paddle/fluid/inference/tensorrt/convert/multiclass_nms_op.cc [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    nvinfer1::Permutation permutation{1, 0};
+    auto* scores_transpose_layer =
+        TRT_ENGINE_ADD_LAYER(engine_, Shuffle, *scores_tensor);
+    scores_transpose_layer->setFirstTranspose(permutation);
+
+    std::vector<nvinfer1::ITensor*> batch_nms_inputs;
+    batch_nms_inputs.push_back(bboxes_expand_layer->getOutput(0));
+    batch_nms_inputs.push_back(scores_transpose_layer->getOutput(0));
+
+    PADDLE_ENFORCE_EQ(engine_->use_oss(), true,
```
Comment (shangzhizhou) [inline paddle/fluid/inference/tensorrt/convert/multiclass_nms_op.cc]: 不应该在这里报错，需要在op_teller里做判断，当功能不支持时，multiclass_nms op不运行在tensorRT里。
在 [OpTeller::Tell()](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/inference/tensorrt/op_teller.cc#L107)时判断是否可以取到需要的plugin来判断？
```
auto creator = GetPluginRegistry()->getPluginCreator("BatchedNMS_TRT", "1");
if (creator == nullptr) return false;
```

File: paddle/fluid/inference/tensorrt/convert/multiclass_nms_op.cc [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+              "Inference multiclass_nms with TensorRT should also "
+              "set parameters %s.",
+              params));
+    };
+    CheckParams(op_desc, "background_label");
+    CheckParams(op_desc, "score_threshold");
+    CheckParams(op_desc, "nms_top_k");
+    CheckParams(op_desc, "nms_threshold");
+    CheckParams(op_desc, "keep_top_k");
+    CheckParams(op_desc, "normalized");
```
Comment (shangzhizhou) [inline paddle/fluid/inference/tensorrt/convert/multiclass_nms_op.cc]: 对这些参数的依赖，需要在 [OpTeller::Tell()](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/inference/tensorrt/op_teller.cc#L107)里做判断，当功能不支持时，multiclass_nms op不运行在tensorRT里。

File: paddle/fluid/inference/tensorrt/convert/multiclass_nms_op.cc (hunk 6 above)
Comment (zlsh80826) [reply paddle/fluid/inference/tensorrt/convert/multiclass_nms_op.cc]: yes, it looks better

**Comment by zlsh80826** [comment]
@shangzhizhou @cryoco, can you help re-run the failed CI checks?

**Comment by shangzhizhou** [comment]
> @shangzhizhou @cryoco, can you help re-run the failed CI checks?

all the required CI are successful， there is no need to rerun the failed ci which is not required，just ignore it.

**Pull request closed without merging**